### PR TITLE
Waves tx json serialization

### PR DESCRIPTION
**Pull request by Tolsi**
## Description

Added Waves transfer transaction json serialization

## Testing instructions

unit test

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [X] Prefix PR title with `[WIP]` if necessary.
- [X] Add tests to cover changes as needed.
- [X] Update documentation as needed.

### CODE REVIEW

File: src/Waves/Transaction.cpp [hunk 1]
```diff
@@ -48,4 +50,26 @@ Data Transaction::serializeToSign() const {
...
+
+    jsonTx["type"] = TransactionType::transfer;
+    jsonTx["version"] = TransactionVersion::V2;
+    jsonTx["fee"] = fee;
+    jsonTx["senderPublicKey"] = Base58::bitcoin.encode(pub_key);
+    jsonTx["timestamp"] = timestamp;
+    jsonTx["proofs"] = json::array({Base58::bitcoin.encode(signature)}).dump();
+    jsonTx["recipient"] = Address(to).string();
+    if (amount_asset != WAVES) {
+        jsonTx["assetId"] = amount_asset;
```
Comment (vikmeup) [inline src/Waves/Transaction.cpp]: `jsonTx["assetId"] = amount_asset;` is this correct, setting amount for assetId

File: src/proto/Waves.proto [hunk 2]
```diff
@@ -18,14 +19,14 @@ message SigningInput {
...
-    bytes public_key = 8;
-
-    bytes private_key = 9;
+    bytes private_key = 8;
 }
 
 // Transaction signing output.
 message SigningOutput {
     bytes signature = 1;
+    string transactionJson = 2;
```
Comment (vikmeup) [inline src/proto/Waves.proto]: can you rename `transactionJson ` => `json`

File: tests/Waves/TransactionTests.cpp [hunk 3]
```diff
@@ -67,4 +70,42 @@ TEST(WavesTransaction, failedSerialize) {
...
+
+    auto signature = Signer::sign(privateKey, tx1);
+
+    auto json = tx1.buildJson(signature);
+
+    ASSERT_EQ(json["type"], TransactionType::transfer);
+    ASSERT_EQ(json["version"], TransactionVersion::V2);
+    ASSERT_EQ(json["fee"], 100000000);
+    ASSERT_EQ(json["senderPublicKey"], "6mA8eQjie53kd4jbZrwL3ZhMBqCX6nzit1k55tR2X7zU");
+    ASSERT_EQ(json["timestamp"], 1526641218066);
```
Comment (vikmeup) [inline tests/Waves/TransactionTests.cpp]: feel free to compare json string in one `ASSERT_EQ`, otherwise lgtm too

Comment (Tolsi) [reply tests/Waves/TransactionTests.cpp]: field order not guaranteed, I think

File: src/Waves/Transaction.cpp (hunk 1 above)
Comment (Tolsi) [reply src/Waves/Transaction.cpp]: yes. I renamed the field so as not to cause confusion

File: tests/Waves/TransactionTests.cpp (hunk 3 above)
Comment (hewigovens) [reply tests/Waves/TransactionTests.cpp]: `json.dump()` will keep the order

Comment (Tolsi) [reply tests/Waves/TransactionTests.cpp]: in the new version, it can not change? Will it be in the order of addition?

**Comment by Tolsi** [comment]
> Can you add json class for gson to android library?

can you link an example of how it should be?

**Pull request merged**